nixos/modules/services/misc/matrix-dendrite.nix

```diff
@@ -0,0 +1,179 @@
+            '';
+          };
+          private_key = lib.mkOption {
+            type = lib.types.nullOr lib.types.path;
+            example = "${workingDir}/matrix_key.pem";
+            default = "${workingDir}/matrix_key.pem";
+            description = ''
+              The path to the signing private key file, used to sign
+              requests and events.
+            '';
+          };
+          trusted_third_party_id_servers = lib.mkOption {
+            type = lib.types.listOf lib.types.str;
+            example = [ "matrix.org" ];
+            default = [ "matrix.org" "vector.im" ];
+            description = ''
+              Lists of domains that the server will trust as identity
+              servers to verify third
+            '';
+          };
+          disable_federation = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Whether to enable federation with other matrix servers.
+            '';
+          };
+        };
+        options.client_api = {
+          registration_disabled = lib.mkOption {
+            type = lib.types.bool;
+            default = false;
+            description = ''
+              Whether to disable user registration to the server
+              without the shared secret.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for dendrite, see:
+        <link xlink:href="https://github.com/matrix-org/dendrite"/>
+      '';
+    };
+  };
+
+  config = lib.mkMerge [
+    (lib.mkIf cfg.enable {
+      systemd.services.matrix-dendrite = {
+        description = "Dendrite Matrix homeserver";
+        after = [
+          "network.target"
+        ] ++ lib.optionals (cfg.generate_private_key == true) [
+          "matrix-dendrite-key-generator.service"
+        ];
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          Type = "notify";
+          DynamicUser = true;
+          StateDirectory = cfg.stateDir;
+          WorkingDirectory = workingDir;
+          ExecStart = lib.strings.concatStringsSep " " ([
+            "${pkgs.matrix-dendrite}/bin/dendrite-monolith-server"
+            "--config ${configurationYaml}"
+            "--http-bind-address :${builtins.toString cfg.httpPort}"
+          ] ++ lib.optionals (cfg.httpsPort != null) [
+            "--https-bind-address :${builtins.toString cfg.httpsPort}"
+          ] ++ lib.optionals (cfg.tls_cert != null) [
+            "--tls-cert ${cfg.tls_cert}"
+          ] ++ lib.optionals (cfg.tls_key != null) [
+            "--tls-key ${cfg.tls_key}"
+          ]);
+          ExecReload = "${pkgs.util-linux}/bin/kill -HUP $MAINPID";
+          Restart = "on-failure";
+        };
+      };
+    }
+    )
+    (lib.mkIf
+      (cfg.enable &&
+        cfg.generate_private_key == true || cfg.tls_cert != null ||
+        cfg.tls_key != null)
```

Comment:
What exactly is the logic behind this condition? And shouldn't it be `cfg.tls_cert == null || cfg.tls_key == null` instead?

Reply:
The reason was that cfg.tls_cert is both the path to the key you want to generate, and the path used to use provided keys. I'm not super happy with how it currently is, and am leaning a bit towards just removing the key generation service.

Reply:
I guess you meant to say
> The reason was that cfg.tls_cert is both the path to the key you want to generate or the one provided by the user

But I'm still very confused about this explanation. How about adding an explanation for how it works to the `tlsCert`, `tlsKey` and/or `generatePrivateKey` options?

I think such a service is valuable if it reduces the need for manual interaction.

Reply:
The generatePrivateKey option reduces the need for manual interaction upon first time setup, I don't think most people will use tlsCert/tslKey because if you do a reverse proxy, this will be handled by nginx.

Would it be excessive to add ad generateTLS option (or generateTLSKey and generateTLSCert options) to make this more clear?

Reply:
I think the changes make it clearer. Let me know what you think.

Reply:
Hmm, doesn't the TLS key and cert have some relation between them? Like, the cert needs to be derived from the key, so it doesn't make any sense to generate a key but have a fixed cert not depending on the key? And is the private key in any way related?

Also, in the current module you'll get a `null` error when setting `generateTls{Key,Cert} = true` but not setting `tls{Key,Cert}` to non-null.

Reply:
Ah, you're right. I simplified the code a bit thank to this, now you have to pass in both. I think you can use your own private_key not related to TLS.

Is there a way to make the option generateTls (the new one I just pushed) depend on tlsKey and tlsCert being non-null? I think the way i have it now, it will silently fail (but not error).

Reply:
How about
```nix
{
  config = {
    services.matrix-dendrite.tlsCert = mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.cert";
    services.matrix-dendrite.tlsKey = mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.key";
  };
}
```

Reply:
Great idea! Added